paddle/fluid/memory/allocation/allocator.h

```diff
@@ -44,13 +46,56 @@ class Allocator;
 // NOTE: this is the base class of Allocation. Each allocator can use its own
 //       allocation object.
 // NOTE: the `Allocation::ptr()` could be nullptr, if the allocation size is 0
+
+/**
+ * Allocation is returned by Allocator::Allocate() method.
+ *
+ * An allocator may be decorated by another allocator. For example, we can
+ * decorate
+ * a RetryAllocator to any allocator to perform allocation retry when first
+ * allocation request fails.
+ *
+ * Explanations of Allocator design is as follows:
+ *
+ * Suppose we have an allocator which is decorated by several allocators:
+ *
+ *   A(1) <- A(2) <- A(3) <- ... <- A(n)
+ *
+ * , and the public allocator is A(1).
+ *
+ * The allocation process would be:
+ *
+ *   A(n).Allocate() -> ... -> A(2).Allocate() -> A(1).Allocate()
```

Comment:
Yes. I am sorry that I have not declared clearly. The actual allocating process would be:

```C++
A(1).Allocate() called -> A(2).Allocate() called -> .... A(n).Allocate() called -> A(n).Allocate() returned -> ... -> A(2).Allocate() returned -> A(1).Allocate() returned
```

The public allocator should call first. But the allocating chain could break if A(i).AllocateImpl() does not call A(i+1).Allocate(). For example, BufferedAllocator would break the allocating chain and return the cached memory.